pkg/lifecycle/api/types.go

```diff
@@ -0,0 +1,51 @@
+package api
+
+import v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// ProductLifecyclePhases holds the lifecycle phases for all versions of all products.
+type ProductLifecyclePhases struct {
+	// ByProduct maps product name to the lifecycle phases for all versions
+	ByProduct map[string]VersionLifecyclePhases `json:"by_product"`
+}
+
+// VersionLifecyclePhases holds the lifecycle phases for all versions of a product.
+type VersionLifecyclePhases struct {
+	// ByVersion maps version to the lifecycle phases
+	ByVersion map[string]LifecyclePhases `json:"by_version"`
+}
+
+// LifecyclePhases holds the lifecycle phases for a version of a product.
+type LifecyclePhases struct {
+	// Previous lists all lifecycle phases that have already occurred. Optional.
+	Previous []LifecyclePhase `json:"previous,omitempty"`
```

Comment:
Note that this differs from the spec - but I believe we need the full list of previous events unless we start inferring that all releases all have the same steps in the same order (which I think is already close to not being the case, e.g. see how we're changing the handling of feature freeze), so this is much safer.

If we don't, then e.g. a config manager which never passes during a phase or which is written after a phase is over may not know that it needs to do something.